### ISSUE TITLE
[terra-button]/[terra-popup] programmatic focus inconsistency during WDIO testing

### ISSUE THREAD

**Issue by PolymorphedSquirrel** [cerner/terra-core#2695]
# Issue Description

During webdriver testing of my application I am seeing reliably inconsistent behavior around the programmatic is-focused class application to the Button.  The functionality seems completely reliable during regular use, but the resulting behavior while undergoing test runs is showing inconsistent application.  Given that our tests are run multiple times (in parallel), the random failure becomes more likely to occur, so getting a successful build and test run via tooling becomes statistically unlikely.

## Additional Context

I have noted that it is possible for the test(s) to pass given enough attempts, so there is possibility of a race condition at some point during the focus application.  I note that focus *is* applied to the button in the screenshot captured and categorized as an error screenshot.  What is missing from this screenshot is the blue halo/outline indicating the focused element which should come from application of the is-focused class to the button.  This began with terra-button 3.20.0 and I have noted increasing unreliability as time passes.

Isolation to a single test being run does not remedy the issue, so this has ruled out test ordering and improper teardown procedures following tests. Minimizing dependency issue possibilities by excluding all non-terra dependencies which could lead to conflicts did not alleviate the inconsistent behavior (though this should not matter due to the behavior of npm allowing each project to pull their own dependencies into themselves within node_modules).

```
browser.waitForVisible('[class*=Button-module__is-focused]', 10000);
```

```
[Terra-Toolkit:serve-static] Server started listening at port:8080

F

0 passing (97.30s)
1 failing

1) My Tests particularTest "before each" hook for "[my_description] to be within the mismatch tolerance":
element ("[class*=Button-module__is-focused]") still not visible after 10000ms
running chrome
Error: element ("[class*=Button-module__is-focused]") still not visible after 10000ms
    at elements("[class*=Button-module__is-focused]") - isVisible.js:54:17
    at isVisible("[class*=Button-module__is-focused]") - waitForVisible.js:73:22
```
terra-button simplified example:
https://gist.github.com/PolymorphedSquirrel/284134cb313a084e8c187c78168e0118

terra-popup simplified example:
https://gist.github.com/PolymorphedSquirrel/2462a7607cee47d4736818e41b7a32c4

This only manifests when running webdriver testing, and only reliably inconsistent when builds are run via Jenkins.  Using a wait with a 10 second timer, 500ms check intervals for the class to be applied to the button, I receive the error given above.  This happens nearly instantly (I cannot detect a delay visually) when manually testing and when using the application, so I believe 10,000ms should be sufficient for even a resource starved environment with a fairly granular test to process the update.

## Expected Behavior
I expect the class application to be consistent when serving the JS locally, in the deployed application, and during webdriver testing.

## Possible Solution
I wish I could offer a possible solution, but I am unsure as to the root cause.

## Environment

* Component Name and Version: terra-button 3.25.0, terra-popup 5.12.0, terra-toolkit 4.29.0
* Browser Name and Version: Chrome (unsure how to grab this version currently)
* Node/npm Version: Node 8.9.2, npm 5.5.1, yarn 1.17.3
* Webpack Version: 4.41.0
* Operating System and version (desktop or mobile): N/A - docker

## @ Mentions
@emilyrohrbough suggested logging an issue for investigation.

**Comment by emilyrohrbough** [comment]
pulling my earlier comment forward:
> I'm not sure about the click behavior on this, or the programatic focus, but I am seeing different behaviors in both your project & terra-ui when click and click off when toggle responsive mode. I also am not seeing focus styles being re-applied when I use escape to close the pop-up. Jankord,Brett do you know the expected behavior with programatic focus?

**Comment by supreethmr** [comment]
@PolymorphedSquirrel can you provide me the link to your tests. so that it will be helpful to find out cause. 
 
@emilyrohrbough I was not able to re-create this Terra-UI and button's class was updated as expected with is-focused class on escape key press in [PopupModalContent](https://engineering.cerner.com/terra-ui/components/terra-popup/popup/popup) example. So would like to if am missing anything.

**Comment by PolymorphedSquirrel** [comment]
@supreethmr I've sent you a message that should be able to provide all necessary information.

**Comment by supreethmr** [comment]
> pulling my earlier comment forward:
> 
> > I'm not sure about the click behavior on this, or the programatic focus, but I am seeing different behaviors in both your project & terra-ui when click and click off when toggle responsive mode. I also am not seeing focus styles being re-applied when I use escape to close the pop-up.

1. `isFocused` class is not being applied for button on outside click when we open toggle responsive mode, It works as expected for outside click in normal mode ( i.e. it sets focus back to button on outside click of popup ) with initial investigation I found that `isFocused` class is being applied to button even in toggle responsive mode but it gets removed on re-render where focused state will have value as false ( need to investigate further on this behaviour  which is happening only on responsive mode).

2. Reason for button not receiving focus on esc key press could be due to Focus-trap Behaviour which currently has issue in setting focus back to previous element when deactivation is triggered via the ESCAPE key. Found a PR logged that has been to fix similar issue on focus-trap : https://github.com/davidtheclark/focus-trap-react/pull/54

And I also observed that we don't have any code return to set focus back to triggered element on close of popup unlike abstract modal which has code written to do something similar to set focus back triggered element on modal close : [code reference](https://github.com/cerner/terra-framework/blob/9fcea5284e7c4b7621b3675249f8d0dadd47e064/packages/terra-abstract-modal/src/AbstractModal.jsx#L159) ) 

In [popup examples](https://engineering.cerner.com/terra-ui/components/terra-popup/popup/popup) focus doesn't seems to be setting back on button onClose of popup in Safari and Mozila unlike Chrome where button is highlighted on close of popup. I am not sure whether this is the issue with Focus-trap behaviour across browser need to investigate further.
